Make CI create comment with link to HTML artifacts

christophebedard opened the pull request:
To make it easier to find the link.

Basically, this:

1. Gets the URL to the artifact from [`actions/upload-artifact`'s output](https://github.com/actions/upload-artifact#outputs)
2. Uses the [`peter-evans/find-comment` action](https://github.com/peter-evans/find-comment) to try to find an existing comment from the bot, if possible
3. Uses the [`peter-evans/create-or-update-comment` action](https://github.com/peter-evans/create-or-update-comment#where-to-find-the-id-of-a-comment) to create a comment and include the artifact URL, or edit the existing comment if it was found

For example, see the comment from the bot below.

Follow-up to #4544

Relates #4548

This should make documenting this "feature" in the ROS 2 docs easier.

christophebedard committed:
Make CI create comment with link to HTML artifacts

Signed-off-by: Christophe Bedard <christophe.bedard@apex.ai>

christophebedard commented:
Let's see if this actually works.

christophebedard commented:
So creating the comment works: https://github.com/ros2/ros2_documentation/actions/runs/9879883370/job/27287129938?pr=4564#step:9:19.

Now I'll amend the commit and force-push to see if it can correctly find and edit the existing comment.

christophebedard commented:
> Now I'll amend the commit and force-push to see if it can correctly find and edit the existing comment.

It works! https://github.com/ros2/ros2_documentation/actions/runs/9879946778/job/27287370231#step:9:20

Now I'll change the content of the comment to include some simple instructions.

kscottz commented:
This is great Christophe! Thanks so much. The Github action looks like it is well maintained so I don't anticipate any issues merging this.

christophebedard committed:
Include PR number in name of HTML arfitacts archive

This way it's easy to know which PR the archive corresponds to, and
this should be less annoying when trying to review multiple PRs at once.

Signed-off-by: Christophe Bedard <christophe.bedard@apex.ai>

kscottz commented:
@christophebedard my intern @akashleena is trying to get #4559 over the line. I propose we get that #4559 over the line, merge this PR, and then @akashleena circles back and do a second PR to fold in your changes. 

I realize that this is a bit convoluted but I think it would be good practice.

christophebedard commented:
Fine with me!

fujitatomoya commented on .github/workflows/test.yml:
```diff
@@ -62,8 +62,30 @@ jobs:
           compression-level: 9 # maximum compression, default 6
           if-no-files-found: error # 'warn' or 'ignore' are also available, defaults to `warn`
+
+      - name: Find HTML artifacts link comment
+        uses: peter-evans/find-comment@v3
+        id: find-comment
+        with:
+          issue-number: ${{ github.event.pull_request.number }}
+          comment-author: 'github-actions[bot]'
+          body-includes: "HTML artifacts:"
+
```
@christophebedard this is nice enhancement! thanks! (thanks Peter Evans too!)

one thing that i want to discuss is, do we want to update the comment or keep posted the comments from action? the reason is, the review history sometimes could be really long, and in that case maybe we want to see the artifacts comment at the bottom of the page after every changes? (in this case, we can see the all artifacts as comments.)

probably you already considered this case, and to avoid many comments from github action, updating the comment on top would be better? I am okay with either way, just wanted to check this.

christophebedard replied:
I considered it, but thought having a single comment would be better. It's probably going to be in the first couple of comments, so at least it's not in the middle of a bunch of comments.

That being said, I don't really mind. I think we could just go with this and switch eventually if we think having separate comments is better.

fujitatomoya reviewed (approved):
lgtm!

christophebedard commented:
Since #4559 has been merged, I'll merge this.

I'll backport this since #4544 was backported too.
